crates/bevy_asset/src/loader_builders.rs

```diff
@@ -25,28 +25,163 @@ impl ReaderRef<'_> {
+///   ([`LoadedUntypedAsset`]).
+///
+///   If you have no idea what type of asset you will be loading (not even at
+///   runtime with a [`TypeId`]), use this.
+///
+/// ## Load mode
+///
+/// To inform the loader how you want to load the asset:
+/// - in [`Deferred`]: when you request to load the asset, you get a [`Handle`]
+///   for it, but the actual loading won't be completed until later.
+///
+///   Use this if you only need a [`Handle`] or [`UntypedHandle`].
+///
+/// - in [`Immediate`]: the load request will load the asset right then and
+///   there, waiting until the asset is fully loaded and giving you access to
+///   it.
+///
+///   Note that this requires you to `await` a future, so you must be in an
+///   async context to use direct loading. In an asset loader, you will be in
+///   an async context.
+///
+///   Use this if you need the *value* of another asset in order to load the
+///   current asset. For example, if you are deriving a new asset from the
+///   referenced asset, or you are building a collection of assets. This will
+///   add the path of the asset as a "load dependency".
+///
+///   If the current loader is used in a [`Process`] "asset preprocessor",
+///   such as a [`LoadTransformAndSave`] preprocessor, changing a "load
+///   dependency" will result in re-processing of the asset.
+///
+/// # Load kickoff
+///
+/// If the current context is a normal [`AssetServer::load`], an actual asset
+/// load will be kicked off immediately, which ensures the load happens as soon
+/// as possible. "Normal loads" kicked from within a normal Bevy App will
+/// generally configure the context to kick off loads immediately.
+///
+/// If the current context is configured to not load dependencies automatically
+/// (ex: [`AssetProcessor`]), a load will not be kicked off automatically. It is
+/// then the calling context's responsibility to begin a load if necessary.
 ///
 /// # Lifetimes
+///
 /// - `ctx`: the lifetime of the associated [`AssetServer`](crate::AssetServer) reference
 /// - `builder`: the lifetime of the temporary builder structs
-pub struct NestedLoader<'ctx, 'builder> {
+///
+/// [`deferred`]: Self::deferred
+/// [`immediate`]: Self::immediate
+/// [`load`]: Self::load
+/// [`with_dynamic_type`]: Self::with_dynamic_type
+/// [`AssetServer::load`]: crate::AssetServer::load
+/// [`AssetProcessor`]: crate::processor::AssetProcessor
+/// [`Process`]: crate::processor::Process
+/// [`LoadTransformAndSave`]: crate::processor::LoadTransformAndSave
+pub struct NestedLoader<'ctx, 'builder, T, M> {
     load_context: &'builder mut LoadContext<'ctx>,
     meta_transform: Option<MetaTransform>,
-    asset_type_id: Option<TypeId>,
+    typing: T,
+    mode: M,
+}
+
+mod sealed {
+    pub trait Typing {}
+
+    pub trait Mode {}
+}
+
+/// [`NestedLoader`] will be provided the type of asset as a type parameter on
+/// [`load`].
+///
+/// [`load`]: NestedLoader::load
+pub struct StaticTyped {
+    _priv: (),
```

Comment:
No real advantage, I like using a struct with named fields this kind of private-only unit struct since the `_priv` makes it super obvious why we need that field.

However, it does mean that initialization has to be spread over multiple lines, which is kind of ugly, so I think I'll change it to a private unit field in a tuple struct.